### PR TITLE
[GDR 1814] Utilize apply_bumpy_function in fit_SE

### PR DESCRIPTION
**Pull request by DominikZabinski**
# Description
## What changed?
Related JIRA issue: 
GDR-1814
## Why was it changed?
Utilize apply_bumpy_function in fit_SE
# Checklist for sustainable code base
- [ ] I added tests for any code changed/added
- [ ] I added documentation for any code changed/added
- [ ] I made sure naming of any new functions is self-explanatory and consistent

# Logistic checklist
- [ ] Package version bumped
- [ ] Changelog updated

# Screenshots (optional)

### CODE REVIEW

File: R/fit_SE.R [hunk 1]
```diff
@@ -15,6 +15,8 @@ fit_SE <- function(se,
 
   # Assertions:
   checkmate::assert_class(se, "SummarizedExperiment")
+  checkmate::assert_string(data_type)
+  checkmate::assert_choice(data_type, c("single-agent", "combination"))
```
Comment (j-smola) [inline R/fit_SE.R:19]: Or just simpler? :) 
```
checkmate::assert_string(data_type, pattern = "single-agent|combination")
```

File: R/fit_SE.R [hunk 2]
```diff
@@ -24,10 +26,8 @@ fit_SE <- function(se,
   checkmate::assert_number(pcutoff)
   checkmate::assert_number(cap)
   checkmate::assert_character(curve_type)
-  
-  req_assays <- c(averaged_assay)
-  lapply(req_assays, function(x) gDRutils::validate_se_assay_name(se, x))
-
+  checkmate::assert_true(all(curve_type %in% c("GR", "RV")))
```
Comment (j-smola) [inline R/fit_SE.R:29]: Or simpler?
```
checkmate::assert_character(curve_type, pattern = "GR|RV")
```

File: R/fit_SE.R [hunk 3]
```diff
@@ -110,3 +76,41 @@ fit_SE <- function(se,
...
+                    nested_identifiers,
+                    n_point_cutoff,
+                    range_conc,
+                    force_fit,
+                    pcutoff,
+                    cap,
+                    curve_type) {
+  fit_df <- S4Vectors::DataFrame(matrix(NA, 2, length(metric_cols)))
+  colnames(fit_df) <- metric_cols
+  rownames(fit_df) <- c("RV", "GR")
```
Comment (j-smola) [inline R/fit_SE.R:96]:
```suggestion
  rownames(fit_df) <- curve_type
```
@gladkia Please, confirm. (scenario: curve_type = "GR")

File: R/fit_SE.R [hunk 4]
```diff
@@ -15,6 +15,8 @@ fit_SE <- function(se,
```
Comment (j-smola) [inline R/fit_SE.R:15]: Maybe description for param could be added?
(description for `gDRutils::fit_curves` may help)

File: R/fit_SE.R (hunk 1 above)
Comment (j-smola) [reply R/fit_SE.R:19]: Other question - is here "combination" correct value?
1. There is a separate function `fit_SE.combination` (for `data_type = "matrix"`)?
2. Param `data_type` is used in `gDRcore::data_model()`.  So if fit_SE is meant for `"single-agent"` type data - `data_type %in% c("single-agent", "cotreatment", "co-dilution")`, right?

File: tests/testthat/test-fit_SE.R [hunk 5]
```diff
@@ -1,9 +1,40 @@
...
+               "'metrics_assay' failed: Must be of type 'string', not 'double'.")
+  expect_error(fit_SE(se, curve_type = 1), 
+               "'curve_type' failed: Must be of type 'character', not 'double'.")
+  expect_error(fit_SE(se, curve_type = c("GR", "dummy")), 
+               "'all(curve_type %in% c(\"GR\", \"RV\"))' failed: Must be TRUE.", 
+               fixed = TRUE)
 })
 
 test_that("fit_SE works as expected", {
+  maeReal <- readRDS(system.file("testdata",
```
Comment (j-smola) [inline tests/testthat/test-fit_SE.R:27]: Since in fit.SE the `data_type` parameter can have two values ​​according to the assertion, shouldn't be also perform test for second?

File: R/fit_SE.R [hunk 6]
```diff
@@ -15,6 +15,8 @@ fit_SE <- function(se,
 
   # Assertions:
   checkmate::assert_class(se, "SummarizedExperiment")
+  checkmate::assert_string(data_type)
+  checkmate::assert_choice(data_type, c("single-agent", "combination"))
   checkmate::assert_character(nested_identifiers, null.ok = TRUE)
   checkmate::assert_string(averaged_assay)
   checkmate::assert_string(metrics_assay)
```
Comment (j-smola) [inline R/fit_SE.R:22]:
```suggestion
```

File: R/fit_SE.R [hunk 7]
```diff
@@ -15,6 +15,8 @@ fit_SE <- function(se,
 
   # Assertions:
   checkmate::assert_class(se, "SummarizedExperiment")
+  checkmate::assert_string(data_type)
+  checkmate::assert_choice(data_type, c("single-agent", "combination"))
   checkmate::assert_character(nested_identifiers, null.ok = TRUE)
   checkmate::assert_string(averaged_assay)
```
Comment (j-smola) [inline R/fit_SE.R:21]:
```suggestion
```

File: R/fit_SE.R [hunk 8]
```diff
@@ -24,10 +26,8 @@ fit_SE <- function(se,
   checkmate::assert_number(pcutoff)
   checkmate::assert_number(cap)
   checkmate::assert_character(curve_type)
-  
-  req_assays <- c(averaged_assay)
-  lapply(req_assays, function(x) gDRutils::validate_se_assay_name(se, x))
-
+  checkmate::assert_true(all(curve_type %in% c("GR", "RV")))
+  gDRutils::validate_se_assay_name(se, averaged_assay)
```
Comment (j-smola) [inline R/fit_SE.R:30]:
```suggestion
  gDRutils::validate_se_assay_name(se, metrics_assay)
```
These both are more strict than
```
  checkmate::assert_string(averaged_assay)
  checkmate::assert_string(metrics_assay)
```

File: tests/testthat/test-fit_SE.R [hunk 9]
```diff
@@ -1,9 +1,40 @@
 test_that("fit_SE errors as expected", {
```
Comment (j-smola) [inline tests/testthat/test-fit_SE.R:1]: Nice check for default values of function's parameters 👍

File: tests/testthat/test-fit_SE.R [hunk 10]
```diff
@@ -1,9 +1,40 @@
...
+               "'all(curve_type %in% c(\"GR\", \"RV\"))' failed: Must be TRUE.", 
+               fixed = TRUE)
 })
 
 test_that("fit_SE works as expected", {
+  maeReal <- readRDS(system.file("testdata", 
+                                 "finalMAE_combo_2dose_nonoise2.RDS", package = "gDRtestData")
+  )
+  
+  nrm_se <- normalize_SE(se = maeReal[[1]], data_type = "single-agent")
```
Comment (j-smola) [inline tests/testthat/test-fit_SE.R:31]: not maeReal[[2]]?
@gladkia

File: tests/testthat/test-fit_SE.R [hunk 11]
```diff
@@ -1,9 +1,40 @@
...
+  expect_error(fit_SE(se = se, data_type = "dummy"), 
+               "'data_type' failed: Must be element of set {'single-agent','combination'}, ", 
+               fixed = TRUE)
+  expect_error(fit_SE(se = se, data_type = "single-agent", nested_identifiers = 1), 
+               "'nested_identifiers' failed: Must be of type 'character' (or 'NULL'), not 'double'", 
+               fixed = TRUE)
+  expect_error(fit_SE(se, averaged_assay = 1), 
+               "'averaged_assay' failed: Must be of type 'string', not 'double'")
+  expect_error(fit_SE(se, averaged_assay = "dummy"), 
+               "'dummy' is not on of the available assays: ''")
```
Comment (j-smola) [inline tests/testthat/test-fit_SE.R:16]: This one is not exactly readable because se has no assay :(
Maybe use
```
se <- <- readRDS(system.file("testdata", 
                                 "finalMAE_combo_2dose_nonoise2.RDS", package = "gDRtestData")
  )[[2]]
```

**Review by j-smola — COMMENTED** [review]
Please, update in DESCRIPTION (becaues of change data.frame -> data.table)
gDRimport (>= 0.99.9)

File: R/fit_SE.R (hunk 1 above)
Comment (DominikZabinski) [reply R/fit_SE.R:19]: First of all I like the idea of using regular expressions over hardcoded values but I feel like mine version is safer and easier to change. For instance `checkmate::assert_string(data_type, pattern = "single-agent|combination")` accepts `data_type = "acombination"` since the pattern checks out but is not a meaningful value of this argument. Writing proper regular expression will be IMO time consuming.
Second of all, I agree with the combination. I will use `gDRcore:::get_data_type_to_data_model_mapping()` function (or `gDRutils::get_experiment_groups()`). Thanks!

File: R/fit_SE.R (hunk 2 above)
Comment (DominikZabinski) [reply R/fit_SE.R:29]: In this case I think mine is easier to maintain over the suggested regular expression.

File: R/fit_SE.R (hunk 8 above)
Comment (DominikZabinski) [reply R/fit_SE.R:30]: averaged_assay: I think it is better to leave both checkmate::assert_string(averaged_assay) and gDRutils::validate_se_assay_name(se, averaged_assay). First checks technical validity, second is for substantive evaluation of the argument.
metrics_assay: I'll add validation on that, thanks!

File: R/fit_SE.R (hunk 7 above)
Comment (DominikZabinski) [reply R/fit_SE.R:21]: I'll stick to this check.

File: R/fit_SE.R (hunk 6 above)
Comment (DominikZabinski) [reply R/fit_SE.R:22]: I'll stick to this check

File: tests/testthat/test-fit_SE.R (hunk 10 above)
Comment (DominikZabinski) [reply tests/testthat/test-fit_SE.R:31]: Good point, I thought I selected single-agent, apparently not. Thanks!

File: tests/testthat/test-fit_SE.R (hunk 11 above)
Comment (DominikZabinski) [reply tests/testthat/test-fit_SE.R:16]: I'll add the tests on non-empty experiment.

File: R/fit_SE.R (hunk 3 above)
Comment (DominikZabinski) [reply R/fit_SE.R:96]: Since gDRutils::fit_curves uses gDRutils:::.apllyLogisticFit and latter apply fitting in specific manner (firstly for RV, secondly for GR) i will change constructing matrix accordingly.

File: tests/testthat/test-fit_SE.R (hunk 5 above)
Comment (DominikZabinski) [reply tests/testthat/test-fit_SE.R:38]: Great point!

Comment (DominikZabinski) [reply tests/testthat/test-fit_SE.R:38]: Great point but for now there is no test data for other data_type than `single-agent`.

File: R/fit_SE.R [hunk 12]
```diff
@@ -110,3 +76,41 @@ fit_SE <- function(se,
   
   se
 }
+
+
+#' @keywords internal
```
[the remark this answers is not in the document]
Comment (DominikZabinski) [reply R/fit_SE.R:81]: Precisely. The goal of this task was to apply `apply_bumpy_function` into fit_SE.

File: R/fit_SE.R (hunk 4 above)
Comment (DominikZabinski) [reply R/fit_SE.R:15]: Since it is a part of runDrugResponseProcessingPipeline, the parameters are described in runDrugResponseProcessingPipelineFxns. There is a separate task for reeavulation of fit_SE.combinations and I suggest reeavulate parameters descritipn for the whole process in that task.

File: tests/testthat/test-fit_SE.R [hunk 13]
```diff
@@ -1,9 +1,58 @@
...
+               "'curve_type' failed: Must be of type 'character', not 'double'.")
+  expect_error(fit_SE(se, curve_type = c("GR", "dummy")), 
+               "'all(curve_type %in% c(\"GR\", \"RV\"))' failed: Must be TRUE.", 
+               fixed = TRUE)
+  
+  maeReal <- readRDS(system.file("testdata", 
+                                 "finalMAE_combo_2dose_nonoise2.RDS", package = "gDRtestData")
+  )
+  
+  ll <- names(maeReal@ExperimentList)
```
Comment (gladkia) [inline tests/testthat/test-fit_SE.R:29]: [minor] consider changing to `experiments(maeReal)`

File: tests/testthat/test-fit_SE.R [hunk 14]
```diff
@@ -1,9 +1,58 @@
...
+  expect_error(fit_SE(se, averaged_assay = "dummy"), 
+               "'dummy' is not on of the available assays")
 })
 
 test_that("fit_SE works as expected", {
+  maeReal <- readRDS(system.file("testdata", 
+                                 "finalMAE_combo_2dose_nonoise2.RDS", package = "gDRtestData")
+  )
+  
+  ll <- names(maeReal@ExperimentList)
```
Comment (gladkia) [inline tests/testthat/test-fit_SE.R:42]: as above

File: tests/testthat/test-fit_SE.R [hunk 15]
```diff
@@ -1,9 +1,58 @@
...
+  se <- maeReal[[ll[which(ll == "single-agent")[1]]]]
+  SummarizedExperiment::assay(se, "Metrics") <- NULL
+  # to avoid warnings about overwriting existing metadata entry
+  S4Vectors::metadata(se)[[".internal"]] <- NULL
```
Comment (gladkia) [inline tests/testthat/test-fit_SE.R:46]: 👍

**Review by gladkia — APPROVED** [review]

File: tests/testthat/test-fit_SE.R (hunk 5 above)
Comment (j-smola) [reply tests/testthat/test-fit_SE.R:37]: @DominikZabinski 
Please add the task - so as not to forget

File: R/fit_SE.R (hunk 8 above)
Comment (j-smola) [reply R/fit_SE.R:30]: ✔️

File: R/fit_SE.R (hunk 2 above)
Comment (j-smola) [reply R/fit_SE.R:29]: .

File: R/fit_SE.R (hunk 6 above)
Comment (j-smola) [reply R/fit_SE.R:22]: .

File: R/fit_SE.R (hunk 7 above)
Comment (j-smola) [reply R/fit_SE.R:21]: .

**Review by j-smola — APPROVED** [review]
